[MM-20158] Show JQL representation of subscription in modal

mickmister opened the pull request:
#### Summary

This PR displays the JQL version of the subscription at the bottom of the modal.

Note that this should always produce valid JQL.

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-20158

#### Screenshots

![image](https://user-images.githubusercontent.com/6913320/68921720-a1645100-0747-11ea-8d77-608c73f82dc8.png)

cpoile commented on webapp/src/utils/jira_issue_metadata.tsx:
```diff
@@ -248,3 +248,84 @@ export function isEpicIssueType(issueType: IssueType): boolean {
+    return s;
+}
+
+export function generateJQLStringFromSubscriptionFilters(issueMetadata: IssueMetadata, fields: FilterField[], filters: ChannelSubscriptionFilters) {
```
One more step towards reimplementing Jira, right? :)

crspeller replied:
I can't wait to have to re-implement this server side for some reason 😆

aaronrothschild commented:
@mickmister Looks good, but could we encapsulate it in a box or something (like a code block).  It looks a bit lost amongst the UI IMO.

asaadmahmood commented:
@aaronrothschild I've updated the styling for the JQL stuff.
![localhost_8065_mattermost_channels_off-topic](https://user-images.githubusercontent.com/11034289/69156934-03d19e80-0b06-11ea-9240-4797df2854a4.png)

DHaussermann commented:
@mickmister I can build and test this now but, when I try the full make, I see the following. Are you able to repro this? There is nothing new available when I try a Git Pull

```
/Users/dylanhaussermann/go/src/github.com/mattermost/mattermost-plugin-jira/webapp/src/utils/styles.js
  15:6  error  Missing semicolon  semi
  16:2  error  Missing semicolon  semi

✖ 2 problems (2 errors, 0 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.


npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! webapp@0.0.1 lint: `eslint --ignore-pattern node_modules --ignore-pattern dist --ignore-pattern tests --ext .js --ext .jsx --ext .ts --ext .tsx . --quiet`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the webapp@0.0.1 lint script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/dylanhaussermann/.npm/_logs/2019-11-19T16_13_15_760Z-debug.log
make: *** [check-style] Error 1
```

DHaussermann commented:
This is working as expected. From tha Mattermost side, I don't see any issues.


But, users will not always be able to take this output of the filters and paste it in Jira to run a search. I'm not sure if that's part of the goal here. There are at least a couple cases where that will not work...

1. If the selected query has labels or field names that match reserved words such as label of `public` 
You will get `Error in the JQL Query: 'Public' is a reserved JQL word. You must surround it in quotation marks to use it in a query. `

2. Queries that contain security level do not work in the UI for me (cloud and server) 
You will get `Field 'Security Level' does not exist or you do not have permission to view it. `

@aaronrothschild is the goal that you can always paste the JQL into the Jira UI?
@mickmister  I'm concerned that the case I found on issue 1. would also affect event delivery. Thjoughts?

mickmister commented:
Instead of shooting for valid JQL, I suggest we have a disclaimer saying that it is an estimation of the resulting query, and that we don't actually use the JQP output. @aaronrothschild What do you think?

> You must surround it in quotation marks to use it in a query

We can do this, but the query looks much less readable with quotes everywhere. And I would rather keep it how it is rather than making a whitelist for things like `public`.

> Field 'Security Level' does not exist or you do not have permission to view it.

It may be called something different in JQL land. Are you able to use the security level field somehow in Jira's JQL search UI? @DHaussermann

mickmister commented:
@DHaussermann lint issues are fixed now

mickmister commented:
Added "Summary" label as Asaad suggested offline. Also added the "Filters" label. No functional changes.

![image](https://user-images.githubusercontent.com/6913320/69408815-af7d1900-0cd5-11ea-91af-549f378ad7b8.png)

aaronrothschild commented:
> Instead of shooting for valid JQL, I suggest we have a disclaimer saying that it is an estimation of the resulting query, and that we don't actually use the JQP output. @aaronrothschild What do you think?
> 
> > You must surround it in quotation marks to use it in a query
> 
> We can do this, but the query looks much less readable with quotes everywhere. And I would rather keep it how it is rather than making a whitelist for things like `public`.
> 
> > Field 'Security Level' does not exist or you do not have permission to view it.
> 
> It may be called something different in JQL land. Are you able to use the security level field somehow in Jira's JQL search UI? @DHaussermann

Sorry, back from vacation now.  

I agree with you @mickmister that it shouldn't be "executable" as is, but should probably have a disclaimer saying "Approximate JQL Equivalent:".  They can paste it in and fix the errors if needed, but I think it provides some extra clarity around what the UI is presenting to the user, not meant to replace the JQL builder (which will always be better than what we can do).
cc @DHaussermann

DHaussermann commented:
@mickmister or @aaronrothschild 
Sorry, it's not clear to me if we need a further change here. Is this `Summary` label instead of a more detailed disclaimer?

mickmister commented:
@DHaussermann How about instead of "Summary", we put "Approximate JQL Output"? Or a caption can be placed below "Summary"

DHaussermann commented:
@mickmister Yes I prefer your suggestion of `Approximate JQL Output`. That seem clear enough to me.

mickmister commented:
@DHaussermann Header has been updated to show "Approximate JQL Output"

@aaronrothschild Do you think this is something we should explain in the documentation?

aaronrothschild commented:
> @aaronrothschild Do you think this is something we should explain in the documentation?

@DHaussermann Yes, I've labeled this for needing documentation so we don't forget.  I'm starting docs tomorrow.
